plugins/hls-alternate-number-format-plugin/src/Ide/Plugin/Conversion.hs

```diff
@@ -0,0 +1,170 @@
+{-# LANGUAGE DeriveGeneric #-}
+module Ide.Plugin.Conversion where
+
+import           Data.Char                     (toUpper)
+import           Data.List                     (delete)
+import           Data.Maybe                    (mapMaybe)
+import           Data.Ratio                    (denominator, numerator)
+import           Data.Text                     (Text)
+import qualified Data.Text                     as T
+import           Development.IDE.Graph.Classes (NFData)
+import           GHC.Generics                  (Generic)
+import           GHC.LanguageExtensions.Type   (Extension (..))
+import           GHC.Show                      (intToDigit)
+import           Ide.Plugin.Literals           (Literal (..), getSrcText)
+import           Numeric
+import           Text.Regex.TDFA               ((=~))
+
+data FormatType = IntFormat IntFormatType
+                | FracFormat FracFormatType
+                | AnyFormat AnyFormatType
+                | NoFormat
+                deriving (Show, Eq, Generic)
+
+instance NFData FormatType
+
+data IntFormatType = HexFormat
+                   | OctalFormat
+                   | BinaryFormat
+                   | NumDecimalFormat
+                   deriving (Show, Eq, Generic)
+
+instance NFData IntFormatType
+
+data FracFormatType = HexFloatFormat
+                    | ExponentFormat
+                    deriving (Show, Eq, Generic)
+
+instance NFData FracFormatType
+
+data AnyFormatType = DecimalFormat
+                   deriving (Show, Eq, Generic)
+
+instance NFData AnyFormatType
+
+alternateFormat :: [FormatType] -> Literal -> [Text]
+alternateFormat fmts lit = case lit of
+  IntLiteral _ (Just _) val  -> concatMap (alternateIntFormat val) (removeCurrentFormat lit fmts)
+  FracLiteral _ (Just _) val -> if denominator val == 1 -- floats that can be integers we can represent as ints
+      then concatMap (alternateIntFormat (numerator val)) (removeCurrentFormat lit fmts)
+      else concatMap (alternateFracFormat val) (removeCurrentFormat lit fmts)
+  _                          -> ["Uh Oh"] -- This means there is no Source Text so we just ignore it
```

Comment:
While I actually chuckled while reading this, we should not use such magic strings to indicate errors.
If this is shown to a user, it is a bad error message.
If this is never shown to a user, it should not be an actual value and rather use `Maybe [Text]`, imo.

Reply:
That should be an empty list, I wanted to make sure I wasn't hitting the rest of the Patterns and forgot about it...